// @matterlabs/hardhat-zksync-deploy/dist/plugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deployLibraries = exports.callDeployScripts = exports.findDeployScripts = void 0;
const fs_1 = require("fs");
const path = __importStar(require("path"));
const fs_2 = __importDefault(require("fs"));
const errors_1 = require("./errors");
const deployer_1 = require("./deployer");
const chalk_1 = __importDefault(require("chalk"));
const utils_1 = require("./utils");
function getAllFiles(dir) {
    const files = [];
    const entries = fs_2.default.readdirSync(dir);
    for (const entry of entries) {
        const entryPath = path.join(dir, entry);
        if (fs_2.default.lstatSync(entryPath).isDirectory()) {
            files.push(...getAllFiles(entryPath));
        }
        else {
            files.push(entryPath);
        }
    }
    return files;
}
function findDeployScripts(hre) {
    const workDir = hre.config.paths.root;
    const deployScriptsDir = path.join(workDir, 'deploy');
    if (!(0, fs_1.existsSync)(deployScriptsDir)) {
        throw new errors_1.ZkSyncDeployPluginError('No deploy folder was found');
    }
    const deployScripts = getAllFiles(deployScriptsDir).filter((file) => path.extname(file) == '.ts' || path.extname(file) == '.js');
    return deployScripts;
}
exports.findDeployScripts = findDeployScripts;
async function callDeployScripts(hre, targetScript) {
    const scripts = findDeployScripts(hre);
    if (targetScript == '') {
        // Target script not specified, run everything.
        for (const script of scripts) {
            await runScript(hre, script);
        }
    }
    else {
        // TODO: Not efficient.
        let found = false;
        for (const script of scripts) {
            if (script.includes(targetScript)) {
                await runScript(hre, script);
                found = true;
                break;
            }
        }
        if (!found) {
            console.error(`Script ${targetScript} was not found, no scripts were run`);
        }
    }
}
exports.callDeployScripts = callDeployScripts;
async function runScript(hre, script) {
    delete require.cache[script];
    let deployFn = require(script);
    if (typeof deployFn.default === 'function') {
        deployFn = deployFn.default;
    }
    if (typeof deployFn !== 'function') {
        throw new errors_1.ZkSyncDeployPluginError('Deploy function does not exist or exported invalidly');
    }
    await deployFn(hre);
}
async function deployLibraries(hre, privateKey, accountNumber, externalConfigObjectPath, exportedConfigObject, noAutoPopulateConfig, compileAllContracts) {
    const wallet = (0, utils_1.getWallet)(hre, privateKey, accountNumber);
    const deployer = new deployer_1.Deployer(hre, wallet);
    const libraryInfos = (0, utils_1.getLibraryInfos)(hre);
    const allDeployedLibraries = [];
    hre.config.zksolc.settings.contractsToCompile = [];
    for (const libraryInfo of libraryInfos) {
        const compileInfo = await deployLibrary(hre, deployer, libraryInfo, libraryInfos, allDeployedLibraries);
        (0, utils_1.fillLibrarySettings)(hre, [compileInfo]);
    }
    console.info(chalk_1.default.green('All libraries deployed successfully!'));
    if (!noAutoPopulateConfig) {
        (0, utils_1.updateHardhatConfigFile)(hre, externalConfigObjectPath, exportedConfigObject);
    }
    (0, utils_1.removeLibraryInfoFile)(hre);
    if (compileAllContracts) {
        console.info(chalk_1.default.yellow('Compiling all contracts'));
        await (0, utils_1.compileContracts)(hre, []);
    }
    else {
        console.info(chalk_1.default.yellow(`Please run ${chalk_1.default.green('yarn hardhat compile')} to compile all contracts`));
    }
}
exports.deployLibraries = deployLibraries;
async function deployLibrary(hre, deployer, missingLibrary, missingLibraries, allDeployedLibraries) {
    const deployedLibrary = allDeployedLibraries
        .find(deployedLibrary => (0, utils_1.generateFullQuailfiedNameString)(missingLibrary)
        .includes((0, utils_1.generateFullQuailfiedNameString)(deployedLibrary.contractFQN)));
    if (deployedLibrary) {
        return deployedLibrary;
    }
    const contractFQN = {
        contractName: missingLibrary.contractName,
        contractPath: missingLibrary.contractPath
    };
    if (missingLibrary.missingLibraries.length == 0) {
        return await compileAndDeploy(hre, deployer, contractFQN, allDeployedLibraries);
    }
    const dependentLibraries = findDependentLibraries(missingLibrary.missingLibraries, missingLibraries);
    const contractInfos = await Promise.all(Array.from(dependentLibraries)
        .map(async (dependentLibrary) => await deployLibrary(hre, deployer, dependentLibrary, missingLibraries, allDeployedLibraries)));
    (0, utils_1.fillLibrarySettings)(hre, contractInfos);
    return await compileAndDeploy(hre, deployer, contractFQN, allDeployedLibraries);
}
function findDependentLibraries(dependentLibraries, missingLibraries) {
    return dependentLibraries.map(dependentLibrary => {
        const dependentFQNString = dependentLibrary.split(':');
        const dependentFQN = {
            contractName: dependentFQNString[1],
            contractPath: dependentFQNString[0]
        };
        const foundMissingLibrary = missingLibraries
            .find(missingLibrary => (0, utils_1.generateFullQuailfiedNameString)(missingLibrary)
            .includes((0, utils_1.generateFullQuailfiedNameString)(dependentFQN)));
        if (!foundMissingLibrary) {
            throw new errors_1.ZkSyncDeployPluginError(`Missing library ${dependentLibrary} not found`);
        }
        return foundMissingLibrary;
    });
}
async function deployOneLibrary(deployer, contractFQN, allDeployedLibraries) {
    const artifact = await deployer.loadArtifact((0, utils_1.generateFullQuailfiedNameString)(contractFQN));
    console.info(chalk_1.default.yellow(`Deploying ${(0, utils_1.generateFullQuailfiedNameString)(contractFQN)} .....`));
    const contract = await deployer.deploy(artifact, []);
    console.info(chalk_1.default.green(`Deployed ${(0, utils_1.generateFullQuailfiedNameString)(contractFQN)} at ${contract.address}`));
    const contractInfo = {
        contractFQN,
        address: contract.address
    };
    allDeployedLibraries.push(contractInfo);
    return contractInfo;
}
async function compileAndDeploy(hre, deployer, contractFQN, allDeployedLibraries) {
    await (0, utils_1.compileContracts)(hre, [contractFQN.contractPath]);
    return await deployOneLibrary(deployer, contractFQN, allDeployedLibraries);
}
